### ISSUE TITLE
createSourceFile build wrong AST in some cases.

### ISSUE THREAD

**Issue by dkulyk** [microsoft/TypeScript#46933]
# Bug Report

### 🕗 Version & Regression Information

- This is a parsing error on 4.5(maybe lower), 4.6-dev

### ⏯ Playground Link

[AST Tree](https://astexplorer.net/#/gist/275f0b786568fff3172a590d680d9184/49214d505f07f8c48b6b2de668ecaa3c21c1cc3f)

### 💻 Code

Parsed JS
```ts repro
!a
    ? (a = (function () {})())
    : (function(a){})(a);
```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?allowJs=true&ts=4.5.2#code/IQQwUABFEPwQFCCBeBAzArgOwMYBcBLAeywQEoIBvAXzPjLMmgC51t9itEya6QyA3EA)

```ts repro
import { createPrinter, createSourceFile, ScriptTarget } from "typescript";

const source = `!a
    ? (a = (function () {})())
    : (function(a){return !0})(a);`

console.log(
    createPrinter().printFile(
        createSourceFile(`test.js`, source, ScriptTarget.Latest)
    )
);
```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?allowJs=true&ts=4.5.2#code/JYWwDg9gTgLgBAbzgYygUwIYzQBSsAO2ygBoV0s0BlCAVymTQDFgAbNMq1YMGAFQxQA5mngBfOADMoEEHABEMAJ5g0AZ2695AbgBQu5BAJr4augzRwAvHAAGAQgy64LuAH44ACgzWvk2gTIMMBGXgCUiGJhnmFhzq4AXH4BQSEE3mEI6DD0BHD2AAxRGdq2+obGEOwAdKwQQp7xLqiY2HiExDHVYPhELOyNrkPkrdTmjP1onrbYJtUAVmq2ZGb0jJya-IIiMNUAMpQmcUNxYXq6QA)

### 🙁 Actual behavior
Wrong JS/TS.
```js
!a
    ? (a = (function () { })()): (function(: a):) => { } : ;
(a);
```
Local tsc for example.ts code make 
```js
"use strict";
!a
    ? (a = (function () { })()) => { } : ;
(a);
```

### 🙂 Expected behavior

```js
!a
    ? (a = (function () { })()): (function (a) { })(a); ;
(a);
```

**Comment by jakebailey** [comment]
Some more reduced examples:

```ts
1 ? (a) : (function() {})         // not OK
1 ? (undefined) : (function() {}) // not OK
1 ? (null) : (function() {})      // OK
1 ? () : (function() {})          // recovers 
```

This appears to be a parser bug (not a printer bug, darn); when the conditional expression gets past the `?`, it tries to parse the first part of `(a) : (function() {})` and the `(a):` tricks it into trying to parse an arrow expression with a missing `=>` (thanks to recovery), rather than just `(a)` alone.

**Comment by jakebailey** [comment]
My fix for this was incomplete; #46962 is a full fix. I didn't consider something like `1 ? (a) : ((((function() {}))))`.
